lib/mauth/rack.rb

```diff
@@ -15,19 +15,23 @@ module Rack
     #   from this is false, the request is passed to the app with no authentication performed.
     class RequestAuthenticator < MAuth::Middleware
       def call(env)
-        if should_authenticate?(env)
-          mauth_request = MAuth::Rack::Request.new(env)
-          begin
-            if mauth_client.authentic?(mauth_request)
-              @app.call(env.merge('mauth.app_uuid' => mauth_request.signature_app_uuid, 'mauth.authentic' => true))
-            else
-              response_for_inauthentic_request(env)
-            end
-          rescue MAuth::UnableToAuthenticateError
-            response_for_unable_to_authenticate(env)
+        mauth_request = MAuth::Rack::Request.new(env)
+        env['mauth.protocol_version'] = mauth_request.protocol_version
+
+        return @app.call(env) unless should_authenticate?(env)
+
+        if mauth_client.v2_only_authenticate? && mauth_request.protocol_version == 1
+          return response_for_missing_v2(env)
+        end
+
+        begin
+          if mauth_client.authentic?(mauth_request)
+            @app.call(env.merge('mauth.app_uuid' => mauth_request.signature_app_uuid, 'mauth.authentic' => true))
```

Comment:
not you, but maybe a slight perf increase with we do an in-place `.merge!` here on 29 or just add a couple lines writing directly to the `env` hash?